dll/win32/browseui/CAutoComplete.cpp

```diff
@@ -25,664 +25,2124 @@
-    AddRef();
+        default:
+            break;
+    }
+    return 0;
+}
 
-    /* If another AutoComplete object was previously assigned to this edit control,
-       release it but keep the same callback on the control, to avoid an infinite
-       recursive loop in ACEditSubclassProc while the property is set to this object */
-    CAutoComplete *prev = static_cast<CAutoComplete *>(GetPropW(m_hwndEdit, autocomplete_propertyW));
+CACEditCtrl::CACEditCtrl() : m_pDropDown(NULL), m_fnOldWordBreakProc(NULL)
+{
+}
 
-    if (prev && prev->m_initialized)
+VOID CACEditCtrl::HookWordBreakProc(BOOL bHook)
+{
+    if (bHook)
     {
-        m_wpOrigEditProc = prev->m_wpOrigEditProc;
-        SetPropW(m_hwndEdit, autocomplete_propertyW, this);
-        prev->m_wpOrigEditProc = NULL;
-        prev->Release();
+        m_fnOldWordBreakProc = reinterpret_cast<EDITWORDBREAKPROCW>(
+            SendMessageW(EM_SETWORDBREAKPROC, 0,
+                reinterpret_cast<LPARAM>(EditWordBreakProcW)));
     }
     else
     {
-        SetPropW(m_hwndEdit, autocomplete_propertyW, (HANDLE)this);
-        m_wpOrigEditProc = (WNDPROC)SetWindowLongPtrW(m_hwndEdit, GWLP_WNDPROC, (LONG_PTR)ACEditSubclassProc);
+        SendMessageW(EM_SETWORDBREAKPROC, 0,
+                     reinterpret_cast<LPARAM>(m_fnOldWordBreakProc));
     }
+}
 
-    if (m_options & ACO_AUTOSUGGEST)
-    {
-        CreateListbox();
-    }
+// WM_CHAR
+// This message is posted to the window with the keyboard focus when WM_KEYDOWN is translated.
+LRESULT CACEditCtrl::OnChar(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &bHandled)
+{
+    TRACE("CACEditCtrl::OnChar(%p, %p)\n", this, wParam);
+    ATLASSERT(m_pDropDown);
+    return m_pDropDown->OnEditChar(wParam, lParam);
+}
 
-    if (pwzsRegKeyPath)
-    {
-        WCHAR *key;
-        WCHAR result[MAX_PATH];
-        WCHAR *value;
-        HKEY hKey = 0;
-        LONG res;
-        LONG len;
+// WM_CLEAR @implemented
+// An application sends this message to an edit control to delete the current selection.
+LRESULT CACEditCtrl::OnClear(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &bHandled)
+{
+    TRACE("CACEditCtrl::OnClear(%p)\n", this);
+    ATLASSERT(m_pDropDown);
+    LRESULT ret = DefWindowProcW(uMsg, wParam, lParam); // do default
+    m_pDropDown->OnEditUpdate(TRUE);
+    return ret;
+}
 
-        /* pwszRegKeyPath contains the key as well as the value, so we split */
-        key = (WCHAR *)HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, (wcslen(pwzsRegKeyPath) + 1) * sizeof(WCHAR));
+// WM_CUT @implemented
+// An application sends this message to an edit control to cut the current selection.
+LRESULT CACEditCtrl::OnCut(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &bHandled)
+{
+    TRACE("CACEditCtrl::OnCut(%p)\n", this);
+    ATLASSERT(m_pDropDown);
+    LRESULT ret = DefWindowProcW(uMsg, wParam, lParam); // do default
+    m_pDropDown->OnEditUpdate(TRUE);
+    return ret;
+}
```

Comment:
This function seems to be identical to the OnClear() one. If they are really meant to be identical, maybe you could group them together? (and both WM_CUT and WM_CLEAR use the same one)

Reply:
United.